ee/desktop/menu/menu.go

```diff
@@ -1,36 +1,186 @@
 )
 
-func Init(hostname string) {
-	onReady := func() {
-		systray.SetTemplateIcon(assets.KolideDesktopIcon, assets.KolideDesktopIcon)
-		systray.SetTooltip("Kolide")
+// MenuData encapsulates a menu bar icon and accessible menu items
+type MenuData struct {
+	Tooltip string         `json:"tooltip,omitempty"`
+	Items   []menuItemData `json:"items"`
+}
 
-		versionItem := systray.AddMenuItem(fmt.Sprintf("Version %s", version.Version().Version), "")
-		versionItem.Disable()
+// menuItemData represents a menu item, optionally containing sub menu items
+type menuItemData struct {
+	Label       string         `json:"label,omitempty"`
+	Tooltip     string         `json:"tooltip,omitempty"`
+	Disabled    bool           `json:"disabled,omitempty"`
+	NonProdOnly bool           `json:"nonProdOnly,omitempty"`
+	IsSeparator bool           `json:"isSeparator,omitempty"`
+	Action      actionData     `json:"action,omitempty"`
+	Items       []menuItemData `json:"items,omitempty"`
+}
 
-		// if prod environment, return
-		if hostname == "k2device-preprod.kolide.com" || hostname == "k2device.kolide.com" {
-			return
-		}
+// MenuBuilder is an interface a menu parser can use to specify how the menu is built
+type MenuBuilder interface {
+	// SetIcon sets the menu bar icon
+	SetIcon()
+	// SetTooltip sets the menu tooltip
+	SetTooltip(tooltip string)
+	// AddMenuItem creates a menu item with the supplied attributes. If the menu item is successfully
+	// created, it is returned. If parent is non-nil, the menu item will be created as a child of parent.
+	AddMenuItem(label, tooltip string, disabled, nonProdOnly bool, action Action, parent any) any
+	// AddSeparator adds a separator to the menu
+	AddSeparator()
+}
+
+type menu struct {
+	logger     log.Logger
+	hostname   string
+	filePath   string
+	buildMutex sync.Mutex
+	doneChans  []chan<- struct{}
+}
+
+func New(logger log.Logger, hostname, filePath string) *menu {
+	m := &menu{
+		logger:   logger,
+		hostname: hostname,
+		filePath: filePath,
+	}
+
+	return m
+}
+
+// Init creates the menu bar & icon. It must be called on the main thread, and
+// blocks until Shutdown() is called.
+func (m *menu) Init() {
+	// Build will be invoked after the menu has been initialized
+	// Before the menu exits, cleanup the goroutines
+	systray.Run(m.Build, m.cleanup)
+}
+
+// Build parses the menu file and constructs the menu. If a menu already exists,
+// all of its items will be removed before the new menu is built.
+func (m *menu) Build() {
+	// Lock so the menu is never being modified by more than one goroutine at a time
+	m.buildMutex.Lock()
+	defer m.buildMutex.Unlock()
+
+	// Remove all menu items each time we rebuild the menu
+	systray.ResetMenu()
+
+	// Even though the menu items have been removed, we still have goroutines hanging around
+	m.cleanup()
+
+	// Reparse the menu file & rebuild the menu
+	menuData := m.getMenuData()
+	if menuData == nil {
+		// TODO: Fall back to default menu?
```

Comment:
What should we do if we can't get the menu from the control server? We can fall back to a simple default menu - maybe just the `Version ...` menu item we have now?

Reply:
Yeah, I think we'll ship a simple static menu. I imagine it will say something like "Waiting to connect to Kolide" and have some basic about info